### DISCUSSION TITLE
AES256-ECB and strange padding

### DISCUSSION THREAD

**Discussion by JKBert** [openssl/openssl#24798] [Q&A] [answered]
Hello,
I'm trying to encrypt and decrypt data between a Linux server and an external device (connected by 4G-LTE). Communication between device and server uses AES256-ECB as it has to run on a very little CPU.
When this device sends data to server, my daemon (written in C) aborts.

Faulty function is this one :
```
unsigned char *
chiffrement(struct_processus *s_etat_processus,
        const EVP_CIPHER *type_chiffrement, logical1 encodage,
        unsigned char *message, integer8 longueur_message,
        unsigned char *clef, integer8 longueur_clef,
        unsigned char *vecteur_initialisation,
        integer8 *longueur_message_chiffre)
{
    int                         longueur_bloc_de_chiffrement;
    int                         longueur_message_1;
    int                         longueur_message_2;

    integer8                    nombre_blocs;

    unsigned char               *message_chiffre;

    EVP_CIPHER_CTX              *contexte;

    if ((contexte = EVP_CIPHER_CTX_new()) == NULL)
    {
        return(NULL);
    }   

    EVP_CIPHER_CTX_reset(contexte);
 
    longueur_bloc_de_chiffrement = EVP_CIPHER_block_size(type_chiffrement);
            
    if (EVP_CipherInit_ex(contexte, type_chiffrement, NULL, clef,
            vecteur_initialisation, (encodage == d_vrai) ? 1 : 0) != 1)
    {
        EVP_CIPHER_CTX_free(contexte);
        return(NULL);
    }

    nombre_blocs = longueur_message / longueur_bloc_de_chiffrement;

    if ((longueur_message % longueur_bloc_de_chiffrement) != 0)
    {
        nombre_blocs++;
    }

    (*longueur_message_chiffre) = nombre_blocs * longueur_bloc_de_chiffrement;
printf("longueur_message_chiffre = %lld\n", (*longueur_message_chiffre));

    // On prévoit une zone de garde pour EVP_CipherFinal_ex() en espérant
    // qu'il ne faille pas plus qu'une longueur de bloc de chiffrement.
    if ((message_chiffre = malloc(((size_t) ((*longueur_message_chiffre)
            + longueur_bloc_de_chiffrement)) *
            sizeof(unsigned char))) == NULL)
    {
        EVP_CIPHER_CTX_free(contexte);
        return(NULL);
    }
printf("longueur_message = %d\n", (int) longueur_message);
    if (EVP_CipherUpdate(contexte, message_chiffre, &longueur_message_1,
            message, (int) longueur_message) != 1)
    {
        free(message_chiffre);
        EVP_CIPHER_CTX_free(contexte);
        return(NULL);
    }
printf("longueur_message_1 = %d\n", longueur_message_1);

    if (EVP_CipherFinal_ex(contexte, message_chiffre + longueur_message_1,
            &longueur_message_2) != 1)
    {
        free(message_chiffre);
        EVP_CIPHER_CTX_free(contexte);
        return(NULL);
    }
printf("longueur_message_2 = %d\n", longueur_message_2);

    (*longueur_message_chiffre) = longueur_message_1 + longueur_message_2;

    // Mise à jour du vecteur d'initialisation

    EVP_CIPHER_CTX_get_updated_iv(contexte, vecteur_initialisation,
            (size_t) EVP_CIPHER_iv_length(type_chiffrement));

    EVP_CIPHER_CTX_free(contexte);

    return(message_chiffre);
}
```

EVP_CipherFinal_ex() returns error. I have done some tests with several different length of messages.

If size of message is 15 bytes :
longueur_message_chiffre = 16
longueur_message = 15
longueur_message_1 = 0
longueur_message_2 = 16

OK, openssl encrypts 15-bytes message in a 16-bytes block.

Now, if size of message is 16 bytes :
longueur_message_chiffre = 16
longueur_message = 16
longueur_message_1 = 16
longueur_message_2 = 16

Thus, I obtain TWO 16 bytes blocks. Why ? One block should be enough. Why openssl does try to form a second encrypted block ?

Data in first block are expected encrypted data. But why a second 16-bytes block ?

Help will be welcome,

JB

**Comment by levitte** [comment]
When cipher padding is enabled, there must *always* be padding (it's checked for consistency when decrypting). Therefore, if the message is a full block, then the padding can only become another full block.

You can control whether you want padding or not, with `EVP_CIPHER_CTX_set_padding()` (you can find it documented [here](https://www.openssl.org/docs/manmaster/man3/EVP_CIPHER_CTX_set_padding.html)). But be aware that if you turn off padding,  you must encrypt an exact number of blocks in total.

**Reply by JKBert** [reply]
Thanks for your answer. But I don't understand. If I encrypt a 16-bytes block (in ECB mode), no padding should be required as each block is independant.

**Reply by levitte** [reply, marked as answer]
So the question to think about is, how will *decryption* know how many bytes (bits, actually) are padding, and how many of them are the actual decrypted message?

If you're interested, https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Padding has a pretty good summary of how that is supposed to work, and the way that's common these days is to add *n* bytes of padding, all having the value *n*.  That way, decryption only has to look at the last byte of the ciphertext, and count backwards that many bytes to know where the plaintext message ends.

To fully grasp it, imagine your message to be encrypted ending with a 15 bytes. In the block to be encrypted, a 1-byte pad will be added, with the value one, so you get something like this:

    { '1', '5', '-', 'b', 'y', 't', 'e', ' ', 'p', 'l', 'n', 't', 'e', 'x', 't', 0x01 }

When the ciphertext of this block gets decrypted, the decryption will see the ending 0x01, and will therefore know that there is 1 byte of padding to be shaved off the end of the decrypted bytes, so the plaintext will become `"15-byte plntext"`, no problem there.

Now, imagine the plaintext to be encrypted like this instead.  I've added a tab (ASCII value 0x09) at the end, so no padding is applied on this block:

    { '1', '5', '-', 'b', 'y', 't', 'e', ' ', 'p', 'l', 'n', 't', 'e', 'x', '\t' }

When the ciphertext of this block gets decrypted, the decryption will see the ending 0x09, and will therefore think that there are 9 byte of padding to be shaved off the end of the decrypted bytes, so the plaintext will become `"15-byt"`...  whoops.

So for this to be decrypted correctly, the encryption will *add* this extra block:

    { 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 
      0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10 }

... and when decrypted, the decryption will see 0x10 and know that there are 16 bytes to be shaved off the end of the decrypted bytes.

-----

Like I said, the other option is to turn padding off, and that it requires that you do padding yourself.

**Reply by JKBert** [reply]
Thanks a lot, I understand better.

**Comment by t8m** [comment]
BTW by using AES256-ECB your implementation has severe security weaknesses. Why aren't you using at least AES256-CTR (of course ensuring that you never reuse the nonce) or even better some efficient AEAD mode such as AES256-OCB?

AES256-CTR is not really slower than AES256-ECB.

**Reply by JKBert** [reply]
I only use ECB as device only has few KB of RAM and only sends a few frames by month. In my case, as device is authenticated by another protocol, ECB is enough.

**Reply by t8m** [reply]
ECB is never enough.

device authenticated != transferred data integrity protected
